File: medium/24. Swap Nodes in Pairs/24. Swap Nodes in Pairs.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next: Optional[ListNode] = next


class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None

        curr_head: Optional[ListNode] = head

        new_head = ListNode(0)
        curr = new_head

        while curr_head and curr_head.next:
            first = curr_head.next.next
            second = curr_head.next

            second.next = curr_head
            curr_head.next = first
            curr.next = second

            curr = curr_head
            curr_head = first

        return new_head.next

File: medium/24. Swap Nodes in Pairs/24. Swap Nodes in Pairs.ts
/**
 * Definition for singly-linked list.
 */

class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

function swapPairs(head: ListNode | null): ListNode | null {
    if (!head || !head.next) return head

    let zero = new ListNode(0, head)

    let first: ListNode | null = head
    let newHead = head.next

    while (first && first.next) {
        const second = first.next
        const three = second.next

        second.next = first
        first.next = three
        zero.next = second

        zero = first
        first = zero.next
    }

    return newHead
};

export {}